include/aos/common/crypto.hpp

```diff
@@ -0,0 +1,253 @@
+     */
+    StaticArray<uint8_t, cCertSubjSize> mSubject;
+    /**
+     * Certificate subject key id.
+     */
+    StaticArray<uint8_t, cCertKeyIdSize> mSubjectKeyId;
+    /**
+     * Certificate authority key id.
+     */
+    StaticArray<uint8_t, cCertKeyIdSize> mAuthorityKeyId;
+    /**
+     * Certificate issuer.
+     */
+    StaticArray<uint8_t, cCertIssuerSize> mIssuer;
+    /**
+     * Certificate serial number.
+     */
+    StaticArray<uint8_t, cCertSerialNumberSize> mSerial;
+    /**
+     * Certificate validity period.
+     */
+    time::Time mNotBefore, mNotAfter;
+};
+
+/**
+ * x509 Certificate request.
+ */
+struct CSR {
+    /**
+     * Raw CSR content.
+     */
+    StaticArray<uint8_t, cRawCSRSize> mRaw;
+    /**
+     * Certificate subject size.
+     */
+    StaticArray<uint8_t, cCertSubjSize> mSubject;
+    /**
+     * Alternative DNS names.
+     */
+    StaticArray<StaticString<cDnsNameLen>, cAltDnsNamesCount> mDnsNames;
+    /**
+     * Contains extra extensions applied to CSR.
+     */
+    StaticArray<asn1::Extension, cCertExtraExtCount> mExtraExtensions;
+};
+
+/**
+ * Provides interface to manage certificate requests.
+ */
+class ProviderItf {
+public:
+    /**
+     * Creates a new certificate based on a template.
+     *
+     * @param templ a pattern for a new certificate.
+     * @param parent a parent certificate in a certificate chain.
+     * @param pubKey public key.
+     * @param privKey private key.
+     * @param[out] resultCert result certificate in PEM format.
+     * @result Error.
+     */
+    virtual Error CreateCertificate(
+        const Certificate& templ, const Certificate& parent, const PrivateKey& privKey, Array<uint8_t>& pemCert)
+        = 0;
+
+    /**
+     * Reads certificates from a PEM blob.
+     *
+     * @param pemBlob raw certificates in a PEM format.
+     * @param[out] resultCerts result certificate chain.
+     * @result Error.
+     */
+    virtual Error PEMToX509Certs(const Array<uint8_t>& pemBlob, Array<Certificate>& resultCerts) = 0;
+
+    /**
+     * Creates a new certificate request, based on a template.
+     *
+     * @param templ template for a new certificate request.
+     * @param privKey private key.
+     * @param[out] resultCSR generated CSR.
+     * @result Error.
+     */
+    virtual Error CreateCSR(const CSR& templ, const PrivateKey& privKey, CSR& resultCSR) = 0;
+
+    /**
+     * Constructs x509 distinguished name(DN) from the argument list.
+     *
+     * @param comName common name part of a result DN.
+     * @param[out] result result DN.
+     * @result Error.
+     */
+    virtual void CreateDN(const String& commonName, Array<uint8_t>& result) = 0;
+
+    /**
+     * Returns text representation of x509 distinguished name(DN).
+     *
+     * @param dn source binary representation of DN.
+     * @param[out] result DN text representation.
+     * @result void.
+     */
+    virtual void DN2String(const Array<uint8_t>& dn, String& result) = 0;
```

Comment:
`DNToString` or rename `PEMToX509Certs` accordingly. `DNToString` is more preferable. We don't use `2` as `To` usually in this project.

It should return err.